packages/clay-data-provider/src/types.ts

```diff
@@ -0,0 +1,182 @@
+	 * Set a value in ms.
+	 */
+	fetchDelay?: number;
+
+	/**
+	 * Options passed to request configuration.
+	 */
+	fetchOptions?: RequestInit;
+
+	/**
+	 * Fetch policy is an option that allows you to specify how you
+	 * want your component to interact with the cache.
+	 *
+	 * (cache-first) Whenever a new request occurs, the data provider
+	 * will first look at the cache and return it if it satisfies the
+	 * data, otherwise it will perform the request.
+	 * (no-cache) It will always make a new request and return the result
+	 * and the cache is deactivated.
+	 * (cache-and-network) This case is specific to when you want your
+	 * users to have a quick response, when a new request happens, it
+	 * will first go to the cache and if it exists it will return and
+	 * regardless if it is in the cache it will make a new request on
+	 * the network.
+	 *
+	 * The data provider takes only the cached data when they meet the
+	 * requirements, the variables and URL are what they define when
+	 * the cache satisfies their request. Be careful if your request
+	 * is needed to cache to avoid problems.
+	 */
+	fetchPolicy?: FetchPolicy;
+
+	/**
+	 * Define the strategies for attempting new requests when it
+	 * fails.
+	 */
+	fetchRetry?: IFetchRetry;
+
+	/**
+	 * Set a request timeout in ms, if it reaches this limit it will
+	 * go through the retry rules and if it still persists, it will
+	 * set the networkStatus to 4 (Error).
+	 */
+	fetchTimeout?: number;
+
+	/**
+	 * Set the URL to where the data provider will have to make a
+	 * request, by default the request is solved with json, if it does
+	 * not cover your use case you can also pass a function by returning
+	 * a Promise. We do not recommend that you use a function to do so,
+	 * you will lose some benefits of the data provider,
+	 * always try to avoid.
+	 */
+	link: TLink;
+
+	/**
+	 * The interval is set in milliseconds, setting the value to zero
+	 * will disable polling.
+	 */
+	pollInterval?: number;
+
+	/**
+	 * Reference your storage provider, like Context, Store Object...
+	 * Whenever a new request happens the data provider will look at
+	 * storage, respecting the fetch policy.
+	 *
+	 * If you have a context serving as your store in your application
+	 * independent of where the user is interacting the data provider
+	 * can retrieve the data from the cache again if it is satisfied.
+	 * The data is removed in LRU (least recently used) order.
+	 */
+	storage?: IStorage;
+
+	/**
+	 * Set the amount of items that can be cached, set to zero will be
+	 * treated as infinite, be aware to set an ideal size to offer a
+	 * positive experience for your user but not use a large amount of memory.
+	 */
+	storageMaxSize?: number;
+
+	/**
+	 * Variables are analyzed and converted to be passed as parameters
+	 * to the query of a GET request, for example.
+	 *
+	 * @example
+	 * {
+	 *   name: 'Matu',
+	 * }
+	 *
+	 * output for the link:
+	 * '/?name=Matu'
+	 */
+	variables?: TVariables;
+}
```

Comment:
After seeing all these types, I'm wondering if we should consider some rules in naming. I noticed we will do `Props`, `FetchPolicy`, `IDataProvider`, `TVariables`. I like the `I**` and `T**` scheme, wonder if it'd be best to just enforce this across clay, which means we would need to change `Props` to `IProps`

Reply:
Yeah I was thinking this today, we're naming our types in several ways and it gets a bit confusing. I also like it, so Props was assuming it was a default, but we can leave `IProps` too.

Reply:
I would keep `FetchPolicy` and `NetworkStatus` if it is an `enum`.

Reply:
A pattern I've seen used pretty successfully is:

- No prefixes by default. Used for things that are "concrete", known types, which is usually what you're dealing with. This keeps thing simple.
- `T` and `T*` for generic types.
- `I*` for interfaces (should be rare).

Reply:
I see that a lot in codebase too. I like to use `I*` for interfaces is that it separates the TypeScript information from a variable with the same name and helps to identify the type. One thing I like about Flow is that we can make it explicit in the import that we are importing the type, I think for these cases it works very well.

```javascript
import type {TypeName} from '...';
```

Reply:
That pattern sounds good to me. I think there are probably some eslint + ts rules we can take advantage of here as well.

Reply:
Oh, it would be nice to add eslint + ts rules to this, is there something we can set for nomenclature? I think we can already add that in #2022.